### PR TITLE
Refactor and extend functionality

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
Since I'd like to use your crate in my projects, I took the liberty to extend some of its functionality.

## Changes

- **No-STD**: Now the crate works in no-std.
- **Extended support**: 12→64 elements (default), up to 128 with `len_128` feature flag
- **Added 1-tuple support** with proper `(T,)` syntax
- **Comprehensive test suite**: `tests/tuple_tests.rs` (sizes 1-10 + edge cases) and `tests/blanket_traits.rs` (trait implementation examples)
- **CI/CD**: Added GitHub Actions workflows for Rust CI and security audits
- **License**: Added MIT license file, which was before only mentioned in the README
- **`typed_tuple_macro`**: Moved macro into an external proc-macro crate
- **New methods**: Added, tested and documented `replace`, `pop`, `swap`, `take`, `split_at` methods
- Added inline directives

In order to implement an index-free trait blanket which depends on the underlying tuple having a given type T, I have also switched from the `const usize` approach to marker structs, since [generic_const_exprs is stil unstable](https://github.com/rust-lang/rust/issues/76560). This gives us the ability to create a `TupleKey` trait which although with some verbosity, allows us to define blankets on tuples.

## Note

Please set `CODECOV_TOKEN` repository secret (free from codecov.io) to enable coverage reporting.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2,36 +2,96 @@
...
+/// assert_eq!((67u8, "Alice", 3.5f64).age(), 67u8);
+/// assert_eq!((15u8, 3.5f64, "Bob").age(), 15u8);
+/// assert_eq!(("Charlie", 56u8, 3.5f64).age(), 56u8);
+/// assert_eq!(("Diana", 4.2f64, 29u8, true).age(), 29u8);
+/// ```
+pub trait TupleKey<Marker> {
+    /// The index of the element associated with the marker type.
+    type Idx;
+}
+
```
Comment (cedtwo) [inline src/lib.rs]: Could you clarify any other use cases for `TupleKey`? I can't see any benefit over the following implementation:
```rust
trait GetAge<const INDEX: usize>: TypedTuple<INDEX, u8> {
    fn age(&self) -> u8 {
        *self.get()
    }
}

impl<const INDEX: usize, T: TypedTuple<INDEX, u8>> GetAge<INDEX> for T {}

#[test]
fn test_get_age() {
    assert_eq!((67u8, "Alice", 3.5f64).age(), 67u8);
    assert_eq!((15u8, 3.5f64, "Bob").age(), 15u8);
    assert_eq!(("Charlie", 56u8, 3.5f64).age(), 56u8);
    assert_eq!(("Diana", 4.2f64, 29u8, true).age(), 29u8);
}
```

File: src/lib.rs [hunk 2]
```diff
@@ -1,71 +1,97 @@
...
 
 /// Trait for tuple element manipulation by type.
-pub trait TypedTuple<const INDEX: usize, T> {
+pub trait TypedTuple<Idx, T> {
+    /// The type of the remaining tuple after popping element of type `T`.
+    type PopOutput;
+    /// The type of the left tuple when splitting at INDEX (inclusive).
+    type SplitLeft;
+    /// The type of the right tuple when splitting at INDEX (inclusive).
+    type SplitRight;
```
Comment (cedtwo) [inline src/lib.rs]: Are both inclusive?

**Review by cedtwo — COMMENTED** [review]
Thanks for contributing. Looks like a good addition to the crate. My main concerns are those mentioned in the comments earlier:
- Can the marker struct be avoided? Are there any other use cases I am not aware of?
- Can all cases of `TypedTuple::<INDEX, _>::method(self)` be called with `tuple.method::<INDEX>()`? My only idea is introducing a new trait as demonstrated earlier.
Please give your thoughts.

File: src/lib.rs (hunk 2 above)
Comment (LucaCappelletti94) [reply src/lib.rs]: That is a copy-paste error, I have now documented it better and even added the `: TypedTuple<Idx, T>` on the split left to really hammer in that the `Idx` is left in the left side. It might be desirable now that I think of it to allow for a couple other methods, where the `Idx` remains on the `right` tuple and another one where the `Idx` is fully extracted out.

**Comment by LucaCappelletti94** [comment]
>     * Can the marker struct be avoided? Are there any other use cases I am not aware of?

At this time const operations are not possible on numbers, but they are possible to a degree with structs.

The most relevant for us is the impossibility to define reverse index operations such as the `TupleKey` trait with a const, such as the following [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=b4b5b490d5e7b311f99d37ae4b8a5e03):

```rust
pub trait TupleKey<M> {
    const INDEX: usize;
}

impl TupleKey<(i32, i64, i64)> for i32 {
    const INDEX: usize = 0;
}

pub trait Get<const INDEX: usize, T> {
    fn get(&self) -> &T;
}

/// This would be implemented for all cases with a macro
impl<T1, T2, T3> Get<0, T1> for (T1, T2, T3) {
    fn get(&self) -> &T1 {
        &self.0
    }
}

pub trait GetInt {
    fn get_int(&self) -> i32;
}

impl<T> GetInt for T
where
    i32: TupleKey<T>,
    T: Get<{<i32 as TupleKey<T>>::INDEX}, i32>
{
    fn get_int(&self) -> i32 {
        *self.get()
    }
}
```

Would get us as error:

```rust
   Compiling playground v0.0.1 (/playground)
error: generic parameters may not be used in const operations
  --> src/lib.rs:27:30
   |
27 |     T: Get<{<i32 as TupleKey<T>>::INDEX}, i32>
   |                              ^ cannot perform const operation using `T`
   |
   = note: type parameters may not be used in const expressions

error: could not compile `playground` (lib) due to 1 previous error
```

When we replace the `const INDEX` with a marker struct in [this other Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=649db74a9f314fedcd4037e9af378b79), the same pattern works:

```rust
pub struct Idx0;

pub trait TupleKey<M> {
    type Idx;
}

impl TupleKey<(i32, i64, i64)> for i32 {
    type Idx = Idx0;
}

pub trait Get<Idx, T> {
    fn get(&self) -> &T;
}

/// This would be implemented for all cases with a macro
impl<T1, T2, T3> Get<Idx0, T1> for (T1, T2, T3) {
    fn get(&self) -> &T1 {
        &self.0
    }
}

pub trait GetInt {
    fn get_int(&self) -> i32;
}

impl<T> GetInt for T
where
    i32: TupleKey<T>,
    T: Get<<i32 as TupleKey<T>>::Idx, i32>
{
    fn get_int(&self) -> i32 {
        *self.get()
    }
}
```

Furthermore, we cannot index `{INDEX + 1}` or `{INDEX - 1}`, but we can define a marker struct (say, `()` representing zero) and a successor operator `S`, and proceed to define all integers with the Peano approach (`S<()>` being 1, `S<S<()>>` being 2, etc).

Additionally, markers allow us to define things like a `Last` index, which we can then associate to the various tuples in a trait to always operate on the last entry in the tuple, and generally speaking all sorts of `smart` indexes.
 
>     * Can all cases of `TypedTuple::<INDEX, _>::method(self)` be called with `tuple.method::<INDEX>()`? My only idea is introducing a new trait as demonstrated earlier.

That is likely to be desirable, we would need to discuss which methods benefit from this ergonomic and which do not.

**Comment by LucaCappelletti94** [comment]
I have now replace `split_at` with four methods to cover four possible cases, since `split_at` was effectively just a rather arbitrary `split_left`  and I am certain there will be cases where the other variants are more desirable.

* `fn split_exclusive(self) -> (Self::SplitLeftExclusive, T, Self::SplitRightExclusive);`
* `fn split_left(self) -> (Self::SplitLeftInclusive, Self::SplitRightExclusive);`
* `fn split_right(self) -> (Self::SplitLeftExclusive, Self::SplitRightInclusive);`
* `fn split_inclusive(self) -> (Self::SplitLeftInclusive, Self::SplitRightInclusive);`
